lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -909,6 +1004,39 @@ static NodeValue simplifyConcatNode(Function *F, ConcatNode *CN) {
     }
   }
 
+  /// If the inputs of concat are all reshape nodes, and the inputs of reshape
+  /// are able to be concatenated, then we can perform concat first and then
+  /// reshape. Concat(Reshape(x) * N) -> Reshape(Concat(x * N)).
+  {
+    llvm::SmallVector<ReshapeNode *, 16> reshapes;
+    for (auto &I : CN->getInputs()) {
+      if (auto *R = dyn_cast<ReshapeNode>(I)) {
+        // The input dims of Reshape nodes should have the same size.
+        if (reshapes.size() == 0 ||
+            reshapes.front()->getInput().dims().size() ==
+                R->getInput().dims().size()) {
+          reshapes.push_back(R);
+          continue;
+        }
+      }
+      // The inputs of Concat have to be all Reshape nodes. All the
+      // Reshape nodes need to also have the same shape.
+      return NodeValue(nullptr);
```

Comment:
I'm a bit concerned that you use returns here. This is just one optimization inside simplifyConcatNode. There are other before it and one could add some further optimizations after it. The optimization should check the pre-conditions, apply the transform and then return the new result if it kicked in. If it does not kick in, it should just pass the control to the code after it.

So, you may need to slightly re-formulate the control flow to follow these rules. May be it is easier just to move almost all of this code into a separate function? And then do something like:

```c++
if (auto transformedConcatNode = tryToOptimizeConcatOfRehapes(N)) {
   return transformedConcatNode;
}
```

Reply:
done